### ISSUE TITLE
Deadlock on dd-trace-processor and dd-task-scheduler

### ISSUE THREAD

**Issue by orf** [DataDog/dd-trace-java#2499]
Hey,
We've experienced the following deadlock related to the datadog trace agent. We're running using OpenJDK 14, specifically the `openjdk:14-buster` docker image. We configure it with gradle like so:

```
    implementation 'com.datadoghq:java-dogstatsd-client:2.11.0'
    implementation group: 'com.datadoghq', name: 'dd-trace-api', version: "+"
    agent group: 'com.datadoghq', name: 'dd-java-agent', version: "+"
```

And construct a `NonBlockingStatsDClientBuilder()` for sending statsd metrics. We only override one specific integration using an environment variable `DD_INTEGRATION_JETTY_ENABLED="true"`

The app is a Kotlin app using http4k. The result of the deadlock is unfortunate: `GET` requests to `/healthcheck` functioned fine, however _all_ other requests time out. That includes `POST` requests to `/healthcheck` or another endpoint.

```
Found one Java-level deadlock:
=============================
"dd-task-scheduler":
  waiting to lock monitor 0x00007f42d8005780 (object 0x00000000ac26f0e0, a com.github.benmanes.caffeine.cache.FSAMS),
  which is held by "dd-trace-processor"
"dd-trace-processor":
  waiting to lock monitor 0x00007f4294008e80 (object 0x00000000836f5da8, a datadog.trace.bootstrap.InternalJarURLHandler$Lock),
  which is held by "dd-task-scheduler"
```

Thread lists:

```
Java stack information for the threads listed above:
===================================================
"dd-task-scheduler":
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.put(BoundedLocalCache.java:2031)
	- waiting to lock <0x00000000ac26f0e0> (a com.github.benmanes.caffeine.cache.FSAMS)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.put(BoundedLocalCache.java:1955)
	at com.github.benmanes.caffeine.cache.LocalManualCache.put(LocalManualCache.java:130)
	at datadog.trace.agent.tooling.CaffeineWeakCache.computeIfAbsent(CaffeineWeakCache.java:44)
	at datadog.trace.agent.tooling.bytebuddy.DDCachingPoolStrategy.typePool(DDCachingPoolStrategy.java:99)
	at datadog.trace.agent.tooling.bytebuddy.DDCachingPoolStrategy.typePool(DDCachingPoolStrategy.java:117)
	at net.bytebuddy.agent.builder.AgentBuilder$ClassFileBufferStrategy$Default$1.typePool(AgentBuilder.java:4587)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:10372)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.access$1600(AgentBuilder.java:10140)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$Java9CapableVmDispatcher.run(AgentBuilder.java:10833)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$Java9CapableVmDispatcher.run(AgentBuilder.java:10771)
	at java.security.AccessController.executePrivileged(java.base@14.0.2/AccessController.java:753)
	at java.security.AccessController.doPrivileged(java.base@14.0.2/AccessController.java:391)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:10330)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$ByteBuddy$ModuleSupport.transform(Unknown Source)
	at sun.instrument.TransformerManager.transform(java.instrument@14.0.2/TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(java.instrument@14.0.2/InstrumentationImpl.java:563)
	at java.lang.ClassLoader.defineClass1(java.base@14.0.2/Native Method)
	at java.lang.ClassLoader.defineClass(java.base@14.0.2/ClassLoader.java:1017)
	at java.security.SecureClassLoader.defineClass(java.base@14.0.2/SecureClassLoader.java:151)
	at java.net.URLClassLoader.defineClass(java.base@14.0.2/URLClassLoader.java:514)
	at java.net.URLClassLoader$1.run(java.base@14.0.2/URLClassLoader.java:422)
	at java.net.URLClassLoader$1.run(java.base@14.0.2/URLClassLoader.java:416)
	at java.security.AccessController.executePrivileged(java.base@14.0.2/AccessController.java:784)
	at java.security.AccessController.doPrivileged(java.base@14.0.2/AccessController.java:691)
	at java.net.URLClassLoader.findClass(java.base@14.0.2/URLClassLoader.java:415)
	at datadog.trace.bootstrap.DatadogClassLoader$DelegateClassLoader.loadClass(DatadogClassLoader.java:175)
	- locked <0x00000000836f5da8> (a datadog.trace.bootstrap.InternalJarURLHandler$Lock)
	at java.lang.ClassLoader.loadClass(java.base@14.0.2/ClassLoader.java:522)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.expireAfterAccessEntries(BoundedLocalCache.java:817)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.expireAfterAccessEntries(BoundedLocalCache.java:801)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.expireEntries(BoundedLocalCache.java:781)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.maintenance(BoundedLocalCache.java:1478)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.performCleanUp(BoundedLocalCache.java:1447)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache$PerformCleanupTask.run(BoundedLocalCache.java:38)
	at datadog.trace.util.AgentTaskScheduler$RunnableTask.run(AgentTaskScheduler.java:40)
	at datadog.trace.util.AgentTaskScheduler$RunnableTask.run(AgentTaskScheduler.java:35)
	at datadog.trace.util.AgentTaskScheduler$PeriodicTask.run(AgentTaskScheduler.java:267)
	at datadog.trace.util.AgentTaskScheduler$Worker.run(AgentTaskScheduler.java:222)
	at java.lang.Thread.run(java.base@14.0.2/Thread.java:832)
"dd-trace-processor":
	at datadog.trace.bootstrap.DatadogClassLoader$DelegateClassLoader.loadClass(DatadogClassLoader.java:169)
	- waiting to lock <0x00000000836f5da8> (a datadog.trace.bootstrap.InternalJarURLHandler$Lock)
	at java.lang.ClassLoader.loadClass(java.base@14.0.2/ClassLoader.java:522)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.put(BoundedLocalCache.java:2042)
	- locked <0x00000000ac26f0e0> (a com.github.benmanes.caffeine.cache.FSAMS)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.put(BoundedLocalCache.java:1955)
	at com.github.benmanes.caffeine.cache.LocalManualCache.put(LocalManualCache.java:130)
	at datadog.trace.agent.tooling.CaffeineWeakCache.computeIfAbsent(CaffeineWeakCache.java:44)
	at datadog.trace.agent.tooling.bytebuddy.DDCachingPoolStrategy.typePool(DDCachingPoolStrategy.java:99)
	at datadog.trace.agent.tooling.bytebuddy.DDCachingPoolStrategy.typePool(DDCachingPoolStrategy.java:117)
	at net.bytebuddy.agent.builder.AgentBuilder$ClassFileBufferStrategy$Default$1.typePool(AgentBuilder.java:4587)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:10372)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.access$1600(AgentBuilder.java:10140)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$Java9CapableVmDispatcher.run(AgentBuilder.java:10833)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$Java9CapableVmDispatcher.run(AgentBuilder.java:10771)
	at java.security.AccessController.executePrivileged(java.base@14.0.2/AccessController.java:753)
	at java.security.AccessController.doPrivileged(java.base@14.0.2/AccessController.java:391)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:10330)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$ByteBuddy$ModuleSupport.transform(Unknown Source)
	at sun.instrument.TransformerManager.transform(java.instrument@14.0.2/TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(java.instrument@14.0.2/InstrumentationImpl.java:563)
	at java.lang.ClassLoader.defineClass1(java.base@14.0.2/Native Method)
	at java.lang.ClassLoader.defineClass(java.base@14.0.2/ClassLoader.java:1017)
	at java.security.SecureClassLoader.defineClass(java.base@14.0.2/SecureClassLoader.java:151)
	at java.net.URLClassLoader.defineClass(java.base@14.0.2/URLClassLoader.java:514)
	at java.net.URLClassLoader$1.run(java.base@14.0.2/URLClassLoader.java:422)
	at java.net.URLClassLoader$1.run(java.base@14.0.2/URLClassLoader.java:416)
	at java.security.AccessController.executePrivileged(java.base@14.0.2/AccessController.java:784)
	at java.security.AccessController.doPrivileged(java.base@14.0.2/AccessController.java:691)
	at java.net.URLClassLoader.findClass(java.base@14.0.2/URLClassLoader.java:415)
	at datadog.trace.bootstrap.DatadogClassLoader$DelegateClassLoader.loadClass(DatadogClassLoader.java:175)
	- locked <0x00000000836f52e0> (a datadog.trace.bootstrap.InternalJarURLHandler$Lock)
	at java.lang.ClassLoader.loadClass(java.base@14.0.2/ClassLoader.java:522)
	at datadog.trace.agent.core.serialization.msgpack.MsgPackWriter.writeString(MsgPackWriter.java:140)
	at datadog.trace.agent.common.writer.ddagent.TraceMapperV0_4$MetaWriter.accept(TraceMapperV0_4.java:92)
	at datadog.trace.agent.core.DDSpanContext.processTagsAndBaggage(DDSpanContext.java:439)
	- locked <0x00000000bd12fe38> (a java.util.HashMap)
	at datadog.trace.agent.core.DDSpan.processTagsAndBaggage(DDSpan.java:432)
	at datadog.trace.agent.common.writer.ddagent.TraceMapperV0_4.map(TraceMapperV0_4.java:148)
	at datadog.trace.agent.common.writer.ddagent.TraceMapperV0_4.map(TraceMapperV0_4.java:22)
	at datadog.trace.agent.core.serialization.msgpack.MsgPackWriter.format(MsgPackWriter.java:80)
	at datadog.trace.agent.common.writer.ddagent.PayloadDispatcher.addTrace(PayloadDispatcher.java:62)
	at datadog.trace.agent.common.writer.ddagent.TraceProcessingWorker$TraceSerializingHandler.onEvent(TraceProcessingWorker.java:152)
	at datadog.trace.agent.common.writer.ddagent.TraceProcessingWorker$TraceSerializingHandler.consumeFromPrimaryQueue(TraceProcessingWorker.java:193)
	at datadog.trace.agent.common.writer.ddagent.TraceProcessingWorker$TraceSerializingHandler.runDutyCycle(TraceProcessingWorker.java:180)
	at datadog.trace.agent.common.writer.ddagent.TraceProcessingWorker$TraceSerializingHandler.run(TraceProcessingWorker.java:169)
	at java.lang.Thread.run(java.base@14.0.2/Thread.java:832)

Found 1 deadlock.
```

**Comment by richardstartin** [comment]
Hi @orf, thanks for the report, which version of the tracer are you using?

**Comment by orf** [comment]
Hey 👋, its resolved to `0.75.0`. Here is a complete breakdown of our runtime dependencies: [dependencies.txt](https://github.com/DataDog/dd-trace-java/files/6152184/dependencies.txt)

**Comment by richardstartin** [comment]
I think #2501 should fix this, but will need try to create a minimal test case to reproduce what I think has happened here.
